### DISCUSSION TITLE
[@rollup/plugin-typescript]  Could not find module 'tslib', which is required by this plugin

### DISCUSSION THREAD

**Discussion by gyhyfj** [rollup/plugins#1533] [Q&A]
Follow the [doc](https://rollupjs.org/command-line-interface/#configplugin-plugin), 
and run `pnpm exec rollup -c --configPlugin typescript `, 
then throw error:
```
[!] (plugin typescript) RollupError: @rollup/plugin-typescript: Could not find module 'tslib', which is required by this plugin. Is it installed?
    at error (D:\GitRepo\rollup-test\node_modules\.pnpm\rollup@3.25.3\node_modules\rollup\dist\shared\rollup.js:349:30)
    at Object.error (D:\GitRepo\rollup-test\node_modules\.pnpm\rollup@3.25.3\node_modules\rollup\dist\shared\rollup.js:1703:20)
    at preflight (D:\GitRepo\rollup-test\node_modules\.pnpm\@rollup+plugin-typescript@11.1.2_rollup@3.25.3_typescript@5.1.6\node_modules\@rollup\plugin-typescript\dist\cjs\index.js:526:17)
    at Object.buildStart (D:\GitRepo\rollup-test\node_modules\.pnpm\@rollup+plugin-typescript@11.1.2_rollup@3.25.3_typescript@5.1.6\node_modules\@rollup\plugin-typescript\dist\cjs\index.js:776:13)
    at D:\GitRepo\rollup-test\node_modules\.pnpm\rollup@3.25.3\node_modules\rollup\dist\shared\rollup.js:1896:40
    at async Promise.all (index 0)
    at PluginDriver.hookParallel (D:\GitRepo\rollup-test\node_modules\.pnpm\rollup@3.25.3\node_modules\rollup\dist\shared\rollup.js:1824:9)
    at D:\GitRepo\rollup-test\node_modules\.pnpm\rollup@3.25.3\node_modules\rollup\dist\shared\rollup.js:26909:13
    at catchUnfinishedHookActions (D:\GitRepo\rollup-test\node_modules\.pnpm\rollup@3.25.3\node_modules\rollup\dist\shared\rollup.js:26167:16)
    at rollupInternal (D:\GitRepo\rollup-test\node_modules\.pnpm\rollup@3.25.3\node_modules\rollup\dist\shared\rollup.js:26906:5)
```

But if I use that typescript2 plugin, this command will work well.

**Comment by tigrr** [comment]
This is a bug. Why was it moved to "Discussion"? And no answer?

**Comment by Yonben** [comment]
Same issue here...

**Comment by zach-betz-hln** [comment]
Am having the same issue with Yarn 4.3.0

edit: just saw another discussion of the same https://github.com/rollup/plugins/discussions/1529

**Comment by MrAmericanMike** [comment]
Same issue here. If @rollup/plugin-typescript needs tslib as a dependency to work properly, then it should install it itself as a dependency and not throw an error.

```
[!] (plugin typescript) RollupError: [plugin typescript] @rollup/plugin-typescript: Could not find module 'tslib', which is required by this plugin. Is it installed?
```

And the answer to the question is clear. Of course it's not installed. Why would it be? I can understand it expecting to find typescript already on a project, but why tslib?

Dependencies of this kind should not be managed by a developer, but by the library authors. Specially because they are the ones that know what version of tslib is needed.

**Reply by shellscape** [reply]
it's a yarn problem. and please don't dictate dependency policy, that's not a good approach. the developer landscape is wider than your machine and preferences.

**Reply by MrAmericanMike** [reply]
Not sure what "It's a yarn problem" means, but as a reference I was using PNPM and not yarn.

PS: Does this mean that this error wouldn't happen if we use NPM?

I have seen PNPM doing weird things, so it wouldn't surprise me.

That been said I see that tslib is listed as a "peerDependency" in the package.json

What's the reason?

I had to do some research on what peerDependencies are, as I was not super familiar with this concept.

```
Peer Dependencies
Definition: These are packages that are expected to be provided by the host application. They indicate that your package is compatible with a specific version of another package but does not directly depend on it.
Installation: When you declare a peer dependency, npm does not install it automatically. Instead, it expects the consumer of your package to install it. This is useful for plugins or libraries that need to work with a specific version of a package (like React).
Usage: You declare them in your package.json under the "peerDependencies" field.
```

"Does not directly depend on it". But in this case it looks like it does, as the library will not work without it.

**Comment by xuhdev** [comment]
The typescript plugin sets `tslib` as an optional peer dependency, which is not installed automatically:

https://github.com/rollup/plugins/blob/92daef00b0da30de172868d4e0792c8686da0045/packages/typescript/package.json#L62-L64

I think a better question here is whether we can improve the error message by prompting the user to install a reasonable version of `tslib`.

**Comment by burner1024** [comment]
Doesn't work for me even with installed tslib.

```
pnpm list rollup tslib @rollup/plugin-typescript typescript
...
dependencies:
@rollup/plugin-typescript 12.1.2
rollup 4.30.1
tslib 2.8.1
typescript 4.9.5
````

`Error [RollupError]: [plugin typescript] @rollup/plugin-typescript: Could not find module 'tslib', which is required by this plugin. Is it installed?`

**Reply by KWangechi** [reply]
@burner1024 Were you able to figure it out?

My solution was to install tslib as a **devDependency** in my project(I'm using npm). With this, the error has been fixed.

But I agree, a better error message would suffice

**Reply by burner1024** [reply]
I figured out something... but don't remember what anymore :).

**Reply by cdfarkas** [reply]
add tslib as devDependencies and as `external` in the rollup config
it did the job for me

**Comment by razzp** [comment]
Still very much an issue over two years later.

**Reply by shellscape** [reply]
You should open a PR to resolve it! Our project relies 100% on community contributions

**Comment by razzp** [comment]
For anyone looking for a quick fix, manually installing `tslib` _does_ work, but it **must** be installed as a dependency. Installing it as a devDependency **will not work**. It seems obvious, but it caught me out.

**Comment by ghost** [comment]
what nuxt 4 use, rolup or rolldow or rollindeep
